default/python/AI/ProductionAI.py

```diff
@@ -1611,59 +1272,57 @@ def already_has_completed_colony_building(planet_id) -> bool:
-    if bld_name in AIDependencies.SYSTEM_SHIP_FACILITIES:
-        current_locs = get_systems_with_facilities(bld_name)
-        current_coverage = current_locs.union(universe.getPlanet(planet_id).systemID for planet_id in queued_bld_locs)
+    queued_bld_pids = building_type.queued_at()
+    if building_type in BuildingType.get_system_ship_facilities():
+        current_coverage = building_type.built_or_queued_at_sys()
         open_systems = set(
-            universe.getPlanet(pid).systemID for pid in get_best_pilot_facilities("BLD_SHIPYARD_BASE")
+            universe.getPlanet(pid).systemID for pid in BuildingType.SHIPYARD_BASE.get_best_pilot_facilities()
         ).difference(current_coverage)
-        try_systems = (
-            open_systems.intersection(get_systems_with_facilities(prereq_bldg)) if prereq_bldg else open_systems
-        )
-        try_locs = set(pid for sys_id in try_systems for pid in get_owned_planets_in_system(sys_id))
+        try_systems = open_systems & prerequisite_type.built_or_queued_at_sys() if prerequisite_type else open_systems
+        try_pids = set(pid for sys_id in try_systems for pid in get_owned_planets_in_system(sys_id))
     else:
-        current_locs = get_best_pilot_facilities(bld_name)
-        try_locs = get_best_pilot_facilities(prereq_bldg).difference(queued_bld_locs, current_locs)
+        current_locs = get_best_pilot_facilities(building_type.value)
+        try_pids = prerequisite_type.get_best_pilot_facilities().difference(queued_bld_pids, current_locs)
     debug(
         "Considering constructing a %s, have %d already built and %d queued",
-        bld_name,
-        len(current_locs),
-        len(queued_bld_locs),
+        building_type,
+        len(building_type.built_at()),
+        len(building_type.queued_at()),
     )
-    max_under_construction = max(1, (time * total_pp) // (5 * this_cost))
-    max_total = max(1, (time * total_pp) // (2 * this_cost))
+    if not try_pids:
+        return
+    # ship facilities all have location independent costs
+    turn_cost = building_type.turn_cost(list(try_pids)[0])
+    max_under_construction = max(1, int(total_pp) // (int(5 * turn_cost)))
+    max_total = max(1, int(total_pp) // int(2 * turn_cost))
     debug("Allowances: max total: %d, max under construction: %d", max_total, max_under_construction)
-    if len(current_locs) >= max_total:
+    if len(building_type.built_at()) >= max_total:
         return
-    valid_locs = list(
-        loc for loc in try_locs.intersection(top_locs) if bld_type.canBeProduced(empire.empireID, loc)
-    ) + list(loc for loc in try_locs.difference(top_locs) if bld_type.canBeProduced(empire.empireID, loc))
-    debug("Have %d potential locations: %s", len(valid_locs), [universe.getPlanet(x) for x in valid_locs])
+    valid_pids = list(pids for pids in try_pids & top_pids if building_type.can_be_produced(pids)) + list(
```

Comment:
This one is quite long, let's split it into chunks. 

`list(pids for pids in try_pids & top_pids if building_type.can_be_produced(pids))`

Please use list comprehansions.